### PR TITLE
HAProxy delegation example

### PR DESCRIPTION
**Pull request by villepeh**
### Pull Request Checklist

Signed-off-by: Ville Petteri Huh.

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)

### CODE REVIEW

**Comment by villepeh** [comment]
HAProxy delegation example for the reverse proxy guide. Based on [talk at Stack Overflow](https://stackoverflow.com/questions/26556838/how-to-send-a-response-with-haproxy-without-passing-the-request-to-web-servers). Requires HAProxy 2.2+

File: docs/reverse_proxy.md [hunk 1]
```diff
@@ -206,6 +206,28 @@ backend matrix
...
+  http-after-response set-header Access-Control-Allow-Origin "*"
+  http-after-response set-header Access-Control-Allow-Methods "GET, POST, PUT, DELETE, OPTIONS"
+  http-after-response set-header Access-Control-Allow-Headers "Origin, X-Requested-With, Content-Type, Accept, Authorization"
+  http-request return status 200 content-type application/json string '{"m.homeserver":{"base_url":"https://matrix.example.com"}}'
+
+backend WKserv
+  http-after-response set-header Access-Control-Allow-Origin "*"
+  http-after-response set-header Access-Control-Allow-Methods "GET, POST, PUT, DELETE, OPTIONS"
+  http-after-response set-header Access-Control-Allow-Headers "Origin, X-Requested-With, Content-Type, Accept, Authorization"
+  http-request return status 200 content-type application/json string '{"m.server":"matrix.example.com:443"}'
```
Comment (squahtx) [inline docs/reverse_proxy.md:228]: For consistency, I think we should:
* follow the naming style of the HAProxy config above
* send the same responses as in the Caddy v2 section above

```suggestion
frontend https
  acl matrix-well-known-client-path path /.well-known/matrix/client
  acl matrix-well-known-server-path path /.well-known/matrix/server
  use_backend matrix-well-known-client if matrix-well-known-client-path
  use_backend matrix-well-known-server if matrix-well-known-server-path
 
backend matrix-well-known-client
  http-after-response set-header Access-Control-Allow-Origin "*"
  http-after-response set-header Access-Control-Allow-Methods "GET, POST, PUT, DELETE, OPTIONS"
  http-after-response set-header Access-Control-Allow-Headers "Origin, X-Requested-With, Content-Type, Accept, Authorization"
  http-request return status 200 content-type application/json string '{"m.homeserver":{"base_url":"https://matrix.example.com"},"m.identity_server":{"base_url":"https://identity.example.com"}}'

backend matrix-well-known-server
  http-after-response set-header Access-Control-Allow-Origin "*"
  http-after-response set-header Access-Control-Allow-Methods "GET, POST, PUT, DELETE, OPTIONS"
  http-after-response set-header Access-Control-Allow-Headers "Origin, X-Requested-With, Content-Type, Accept, Authorization"
  http-request return status 200 content-type application/json string '{"m.server":"matrix.example.com:443"}'
```

File: changelog.d/12501.doc [hunk 2]
```diff
@@ -0,0 +1 @@
+HAProxy delegation example with CORS headers.
```
Comment (squahtx) [inline changelog.d/12501.doc:1]: The changelog entry could use a verb, eg.
```suggestion
Add HAProxy delegation example with CORS headers to docs.
```

**Review by squahtx — COMMENTED** [review]
Thanks for updating the docs! Looks good, just some minor suggestions.

**Comment by villepeh** [comment]
Thank you very much! These edits definitely make it way more neat and easier to read.

**Review by squahtx — APPROVED** [review]
LGTM! Thanks for contributing to the docs!

**Pull request merged**